### PR TITLE
HPCC-19949 Persist_replicate.ecl failed on Roxie after "make install".

### PR DESCRIPTION
**Pull request by AttilaVamos**
Update mspec.defaultBaseDir and mspec.defaultReplicateDir if they are
differ from the group default base directory

Signed-off-by: HPCCSmoketest <hpccsmoketest@gmail.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:

Tested manually in HPCC system generated with make install.

### CODE REVIEW

**Comment by AttilaVamos** [comment]
@jakesmith please review.

File: dali/base/dafdesc.cpp [hunk 1]
```diff
@@ -432,11 +432,14 @@ struct CClusterInfo: implements IClusterInfo, public CInterface
...
-            if (mspec.defaultBaseDir.isEmpty())
+            bool baseDirChanged = false;
+            if (mspec.defaultBaseDir.isEmpty() ||
+               (!defaultDir.isEmpty() && (strcmp(mspec.defaultBaseDir.str(),defaultDir.str()) != 0)) )
             {
                 mspec.setDefaultBaseDir(defaultDir);   // MORE - should possibly set up the rest of the mspec info from the group info here
+                baseDirChanged = true;
             }
-            if (mspec.defaultCopies>1 && mspec.defaultReplicateDir.isEmpty())
+            if ((mspec.defaultCopies>1 && mspec.defaultReplicateDir.isEmpty()) || baseDirChanged)
```
Comment (jakesmith) [inline dali/base/dafdesc.cpp]: Shouldn't this be:
```
if (mspec.defaultCopies>1 && (mspec.defaultReplicateDir.isEmpty() || baseDirChanged))
```

i.e. shouldn't be setting replicate dir if <=1 copies ..

**Review by jakesmith — CHANGES_REQUESTED** [review]

**Comment by jakesmith** [comment]
@AttilaVamos - please see comments.

File: dali/base/dafdesc.cpp [hunk 2]
```diff
@@ -432,11 +432,14 @@ struct CClusterInfo: implements IClusterInfo, public CInterface
             GroupType groupType;
             group.setown(resolver->lookup(name.get(), defaultDir, groupType));
             // MORE - common some of this with checkClusterName?
-            if (mspec.defaultBaseDir.isEmpty())
+            bool baseDirChanged = false;
+            if (mspec.defaultBaseDir.isEmpty() ||
+               (!defaultDir.isEmpty() && (strcmp(mspec.defaultBaseDir.str(),defaultDir.str()) != 0)) )
```
[the remark this answers is not in the document]
Comment (AttilaVamos) [reply dali/base/dafdesc.cpp]: Changed

File: dali/base/dafdesc.cpp (hunk 1 above)
Comment (AttilaVamos) [reply dali/base/dafdesc.cpp]: Changed

**Comment by AttilaVamos** [comment]
@jakesmith updated, please check again.

**Comment by jakesmith** [comment]
@AttilaVamos - looks good.

@richardkchapman - ready to merge

**Comment by richardkchapman** [comment]
@AttilaVamos Please squash

**Comment by AttilaVamos** [comment]
@richardkchapman done

**Pull request merged**